File: lib/src/repository/ride_repository.dart
import '../models/ride.dart';

/// Repository interface for ride-related operations
abstract class RideRepository {
  /// Schedules a new ride
  Future<Ride> scheduleRide(Ride ride);

  /// Schedules a round-trip (two separate rides)
  Future<Ride> scheduleRoundTrip({
    required DateTime pickup,
    required DateTime dropoff,
    required String pickupAddress,
    required String destinationAddress,
    required String notes,
  });

  /// Gets a ride by id
  Future<Ride> getRide(String rideId);

  /// Cancels a ride
  Future<bool> cancelRide(String rideId);

  /// Updates details of an existing ride
  Future<Ride> updateRideDetails(Ride ride);

  /// Gets all upcoming rides
  Future<List<Ride>> getUpcomingRides();

  /// Gets all past rides
  Future<List<Ride>> getPastRides();

  /// Searches rides by date range
  Future<List<Ride>> searchRidesByDate(DateTime startDate, DateTime endDate);

  /// Observes changes to upcoming rides
  Stream<List<Ride>> observeUpcomingRides();

  /// Observes changes to a specific ride
  Stream<Ride?> observeRide(String rideId);
}

File: data/repository/RideRepository.kt
package com.example.healthride.data.repository

import com.example.healthride.data.model.Ride
import kotlinx.coroutines.flow.Flow
import java.util.Date

interface RideRepository {
    // Ride scheduling operations
    suspend fun scheduleRide(ride: Ride): Result<Ride>
    suspend fun scheduleRoundTrip(pickup: Date, dropoff: Date, addresses: Pair<String, String>, notes: String): Result<Ride>

    // Ride management
    suspend fun getRide(rideId: String): Result<Ride>
    suspend fun cancelRide(rideId: String): Result<Boolean>
    suspend fun updateRideDetails(ride: Ride): Result<Ride>

    // Ride retrieval
    suspend fun getUpcomingRides(): Result<List<Ride>>
    suspend fun getPastRides(): Result<List<Ride>>
    suspend fun searchRidesByDate(startDate: Date, endDate: Date): Result<List<Ride>>

    // Observing data changes
    fun observeUpcomingRides(): Flow<List<Ride>>
    fun observeRide(rideId: String): Flow<Ride?>
}